pkg/performanceprofile/cmd/render/render.go

```diff
@@ -143,22 +209,102 @@ func render(inputDir, outputDir string) error {
 		partitioningMode = &infra.Status.CPUPartitioning
 	}
 
-	if isLegacySNOWorkloadPinningMethod(mcConfigs, infra, partitioningMode) {
-		legacyAllNodes := apicfgv1.CPUPartitioningAllNodes
-		partitioningMode = &legacyAllNodes
-	}
-
-	if err := genBootstrapWorkloadPinningManifests(partitioningMode, outputDir, defaultMCPNames...); err != nil {
-		return err
-	}
+	if renderTune {
+		if len(tune) == 0 {
+			klog.Warning("zero tune patch were found")
+		}
+		if isLegacySNOWorkloadPinningMethod(mcConfigs, infra, partitioningMode) {
+			legacyAllNodes := apicfgv1.CPUPartitioningAllNodes
+			partitioningMode = &legacyAllNodes
+		}
 
-	// If the user supplies extra machine pools, we ingest them here
-	for _, pool := range mcPools {
-		if err := genBootstrapWorkloadPinningManifests(partitioningMode, outputDir, pool.Name); err != nil {
+		if err := genBootstrapWorkloadPinningManifests(partitioningMode, outputDir, defaultMCPNames...); err != nil {
 			return err
 		}
-	}
 
+		// If the user supplies extra machine pools, we ingest them here
+		for _, pool := range mcPools {
+			if err := genBootstrapWorkloadPinningManifests(partitioningMode, outputDir, pool.Name); err != nil {
+				return err
+			}
+		}
+
+		for _, tt := range tune {
+			var kernelArguments []string
+
+			mc := tt.Spec.Recommend[0].MachineConfigLabels["machineconfiguration.openshift.io/role"]
+			var mcp *mcfgv1.MachineConfigPool
+			for _, pool := range mcPools {
+				if pool.Name == mc {
+					mcp = pool
+				}
+			}
+			if mcp == nil {
+				klog.Infof("render: No MachineConfigPool found for Tune patch %s", tt.Name)
+				continue
+			}
+
+			tunedData := []byte(*tt.Spec.Profile[0].Data)
+			cfg, err := ini.Load(tunedData)
+			if err != nil {
+				klog.Infof("Failed to load tune data (%s)", *tt.Spec.Profile[0].Data)
+				continue
+			}
+
+			mainSection, err := cfg.GetSection("main")
+			if err != err {
+				klog.Infof("No main section in tune data (%s)", tunedData)
+				continue
+			}
+
+			profileName := mainSection.Key("include").String()
```

Comment:
do we need to read included tuned profiles? I thought `tuned` would do that.

Reply:
The idea is to render performanceProfile and Tuned patch (both part of the DU profile). FOr example:
```yaml
apiVersion: tuned.openshift.io/v1
kind: Tuned
metadata:
  name: performance-patch
  namespace: openshift-cluster-node-tuning-operator
spec:
  profile:
  - data: |
      [main]
      summary=Configuration changes profile inherited from performance created tuned
      include=openshift-node-performance-openshift-node-performance-profile
      [bootloader]
      cmdline_crash=nohz_full=4-47
      [sysctl]
      kernel.timer_migration=1
      [scheduler]
      group.ice-ptp=0:f:10:*:ice-ptp.*
      [service]
      service.stalld=start,enable
      service.chronyd=stop,disable
    name: performance-patch
  recommend:
  - machineConfigLabels:
      machineconfiguration.openshift.io/role: master
    priority: 19
    profile: performance-patch
---
apiVersion: performance.openshift.io/v2
kind: PerformanceProfile
metadata:
  name: openshift-node-performance-profile
spec:
  additionalKernelArgs:
  - rcupdate.rcu_normal_after_boot=0
  - efi=runtime
  - vfio_pci.enable_sriov=1
  - vfio_pci.disable_idle_d3=1
  - module_blacklist=irdma
  cpu:
    isolated: 4-47
    reserved: 0-3
  hugepages:
    defaultHugepagesSize: 1G
    pages:
    - count: 32
      size: 1G
  machineConfigPoolSelector:
    pools.operator.machineconfiguration.openshift.io/master: ""
  nodeSelector:
    node-role.kubernetes.io/master: ""
  numa:
    topologyPolicy: restricted
  realTimeKernel:
    enabled: true
  workloadHints:
    highPowerConsumption: false
    perPodPowerManagement: false
    realTime: true

```
I'm not sure how to do that with `Tuned` - do we need to extract this patch to some file before invoking the `Tuned`?

Reply:
In my understanding the target is to create a MachineConfig to apply kernerl boot arguments.
To get those arguments we need to run tuned with the `tuned.conf` file generated by the PerformanceProfile and then read the `/etc/tuned/bootcmdline` file.

So, as far as I see it, we could use `PerformanceProfile` as "main" input and use [NewNodePerformance function](https://github.com/openshift/cluster-node-tuning-operator/blob/12b9a12c1cf6892b36464c26d6198d2211f28a14/pkg/performanceprofile/controller/performanceprofile/components/tuned/tuned.go#L55) to get the `tuned` from there, as the PerformanceProfile controller does.

Other inputs should be the MachineConfigPools for the SNO ( we already have them cause they are used in the previous step of the installer, the one for the PerformanceProfile render command [here](https://github.com/openshift/installer/blob/b36d7c201e0114aa0e9261cf8ae5b8d68c74bf07/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L297)

All those inputs will be in one folder (INPUT_DIR)

With all that in place `tuned` will be run and will generate the `/etc/tuned/bootcmdline` file that could be read using [getBootCmdline function](https://github.com/openshift/cluster-node-tuning-operator/blob/12b9a12c1cf6892b36464c26d6198d2211f28a14/pkg/tuned/controller.go#L748)
Here the challenge is to mount all the system folders `tuned` needs to work properly from the host machine but without letting `tuned` to modify them.  I am working with overlayfs and `podman-run` to get that.

With all that in place we just need to create the MachineConfig object and write the manifest to a folder which should be mounted from the host machine so it can be read after the container finish its execution.

This is just my understanding of the task, so maybe @MarSik or @yanirq can correct me.

A (WIP) implementation of this is https://github.com/openshift/cluster-node-tuning-operator/pull/844

Reply:
You probably need to read all extra Tuned objects too. Customers can use those to override pieces of our default one (https://access.redhat.com/solutions/5532341).

But the rest sounds right to me:

Prepare the (RHEL level) tuned profiles, give them to tuned, run the oneshot mode (in isolation somehow), collect the boot args, generate MC compatible with what NTO would do at Day 2.